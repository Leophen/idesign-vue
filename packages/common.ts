import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
