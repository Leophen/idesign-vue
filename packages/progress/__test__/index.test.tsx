import { Progress } from '../index';
import { render } from '../../common'

describe('Progress ç»„ä»¶æµ‹è¯•', () => {
  it('create', () => {
    const wrapper = render(<Progress percentage={60} />);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('label', () => {
    const wrapper = render(
      <Progress
        v-slots={{
          label: () => "ğŸ˜„"
        }}
      />
    );
    expect(wrapper.find('.i-progress__info').text()).toBe('ğŸ˜„');
  });

  it('hide label', () => {
    const wrapper = render(<Progress labelTxt={false} />);
    expect(wrapper.find('.i-progress__info').exists()).toBe(false);
  });

  it('innerLabel', () => {
    const wrapper = render(<Progress innerLabel />);
    expect(wrapper.find('.i-progress-bar').classes()).toContain('i-progress-bar__has-label');
  });

  it('color backColor', () => {
    const wrapper = render(<Progress color="blue" backColor="red" />);
    expect(wrapper.find('.i-progress-bar').attributes('style')).toMatch('background: red');
    expect(wrapper.find('.i-progress-bar__inner').attributes('style')).toMatch('background: blue');
  });

  it('width', () => {
    const wrapper = render(<Progress width={300} />);
    expect(wrapper.find('.i-progress-bar').attributes('style')).toMatch('width: 300px');
  });

  it('strokeWidth', () => {
    const wrapper = render(<Progress strokeWidth={30} />);
    expect(wrapper.find('.i-progress-bar').attributes('style')).toMatch('height: 30px');
  });

  it('circle', () => {
    const wrapper = render(<Progress type="circle" percentage={60} width={100} strokeWidth={20} />);
    expect(wrapper.element).toMatchSnapshot();
  });
});

