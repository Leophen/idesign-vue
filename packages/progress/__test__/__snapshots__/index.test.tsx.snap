// Vitest Snapshot v1

exports[`Progress 组件测试 > circle 1`] = `
<div
  class="i-progress"
>
  <div
    class="i-progress-circle"
    style="width: 100px; height: 100px;"
  >
    <svg
      height="100"
      width="100"
    >
      <circle
        class="i-progress-circle__bg"
        cx="50"
        cy="50"
        fill="none"
        r="40"
        style="stroke-width: 20px;"
      />
      <circle
        class="i-progress-circle__inner"
        cx="50"
        cy="50"
        fill="none"
        r="40"
        style="stroke-width: 20px; stroke-linecap: round; stroke-dasharray: 314.1592653589793; stroke-dashoffset: 125.66370614359175; animation: unset;"
      />
    </svg>
    <div
      class="i-progress__info"
    >
      <!--v-if-->
       60%
    </div>
  </div>
</div>
`;

exports[`Progress 组件测试 > create 1`] = `
<div
  class="i-progress"
>
  
  <div
    class="i-progress-bar"
    style="width: 200px; height: 6px;"
  >
    <div
      class="i-progress-bar__inner"
      style="width: 60%; animation: unset;"
    />
    <!--v-if-->
  </div>
  <div
    class="i-progress__info"
  >
    60% 
    <!--v-if-->
  </div>
  
</div>
`;
